[enhancement] store balances at block heights - implementation

joel-u410 opened the pull request:
Pushing this as draft, in order to show my proposed changes here. This cannot be merged until the corresponding feature is merged in the SDK: https://github.com/anoma/namada/pull/3530. After that's done, I'll come back here and update this to use the upstream SDK and not my fork.

This implements https://github.com/anoma/namada-indexer/issues/77.

Example of the `balance_changes` table:
![image](https://github.com/user-attachments/assets/cf5ae18a-aac9-4a1b-a973-0f62558f2730)

And corresponding `balances` view (which maintains the previous behavior of the `balances` table):
![image](https://github.com/user-attachments/assets/ea41f6cf-0ba3-46f6-acfa-e60aefdcea49)

After a transaction occurs which updates the balance (note the `id` changes in the returned row in the `balances` view to reflect which balance is now the latest one):
![image](https://github.com/user-attachments/assets/eb625427-aea0-46d9-8dc5-fcd48268fe4c)

joel-u410 commented:
Good news -- https://github.com/anoma/namada/pull/3530 was merged, so I suppose after the next `namada` SDK release, this can be updated / rebased and will no longer require a forked version.

mateuszjasiuk commented:
@joel-u410 👋 Do you still need this? I think your PR to namada was merged some time ago

joel-u410 commented:
> @joel-u410 👋 Do you still need this? I think your PR to namada was merged some time ago

Hi @mateuszjasiuk yes I'd still like this change to get merged (though it is out of date and needs some work now!)

I just haven't had the chance to come back to this project yet (have been waiting until closer to mainnet launch to refocus here). Please keep it open, if it's not too much trouble, and I will update/rebase this soon (probably next week) based on the current namada SDK which does have my PR merged now. Thanks!

mateuszjasiuk commented:
> > @joel-u410 👋 Do you still need this? I think your PR to namada was merged some time ago
> 
> Hi @mateuszjasiuk yes I'd still like this change to get merged (though it is out of date and needs some work now!)
> 
> I just haven't had the chance to come back to this project yet (have been waiting until closer to mainnet launch to refocus here). Please keep it open, if it's not too much trouble, and I will update/rebase this soon (probably next week) based on the current namada SDK which does have my PR merged now. Thanks!

Sure, no problem!

joel-u410 commented:
@mateuszjasiuk I've rebased this now + resolved conflicts on latest `main` -- should be ready to go! 🙏 

I was not able to run `cargo test` locally (it gave errors about the test database not existing) -- is there something I'm missing there? (I used `docker-compose-db.yml` to run a local database and set `DATABASE_URL_TEST` to point to it).

Would like to make sure all tests are passing before this is merged.

joel-u410 commented:
re: running `cargo test` -- I created a database `namada-indexer-test` and my `DATABASE_URL_TEST` points to it. I get errors like this:

```
    0: Failed to establish connection: connection to server at "0.0.0.0", port 5433 failed: FATAL:  database "namada-indexer-test/test_db_1535_21" does not exist
    1: connection to server at "0.0.0.0", port 5433 failed: FATAL:  database "namada-indexer-test/test_db_1535_21" does not exist
```

It seems like I'm missing something here where the tests are supposed to have a new database for each test run?

mateuszjasiuk commented:
You can try this: `DATABASE_URL_TEST=postgres://postgres:password@0.0.0.0:5433 cargo test` I'm not sure why cargo test is not picking up env. I did not have time to check it out 😅

joel-u410 commented:
ah, I see ... `DATABASE_URL_TEST` must not specify a database. Thanks, that worked.

joel-u410 commented:
ok, so there is one failing test on this branch. I'll update & make sure that's passing before marking this as ready.

joel-u410 commented:
@mateuszjasiuk this is ready for review now. I updated the test `test_insert_balance_with_conflicting_owners` to be more correct, I think, based on making it more like `test_insert_balance_with_conflicting_tokens` -- as it was written, it was actually just inserting an updated balance at the same height for the same owner + token, which (rightfully) fails due to the uniqueness constraint in the database.

I also added another test case to test the behavior when a different balance amount is inserted with the same `owner`, `token`, _and_ `height` (the new balance in that case is ignored).

mateuszjasiuk commented on chain/src/repository/balance.rs:
```diff
@@ -21,21 +18,29 @@ pub fn insert_balance(
+    tracing::info!("Inserting {} balances into db", balances.len());
+
+    // Group balances into chunks to avoid hitting the limit of the number of bind parameters in one query.
+    balances.chunks(10000).try_for_each(|balances_chunk| {
```
I'm not sure if it fits your usecase but we also have `insert_balance_in_chunks` right below this function

joel-u410 replied:
Ah, somehow missed that -- I'll give it a try!

joel-u410 replied:
Took a look at `insert_balance_in_chunks`, and it's perfect for my use case here.

Separately, I've made an improvement to get the column count at compile-time instead of querying `information_schema`. I'll send that in a separate PR.

mateuszjasiuk commented on chain/src/repository/balance.rs:
```diff
@@ -216,19 +231,32 @@ mod tests {
         db.run_test(move |conn| {
             seed_balance(conn, vec![balance.clone()])?;
 
+            // TODO: this is probably not a valid way to construct an IbcToken
```
I think it's fine to remove TODO but leave comment

mateuszjasiuk reviewed (commented):
Tested and I think this looks great. Left couple of comments for discussion, but in general awesome work!

EDIT: also my small concern is that up until this point we would just store current state, so if someone starts indexing late then the balance changes will start at block x not first one. If we provide new endpoint to get balance at height we probably need to mention this in the description

Fraccaman commented on orm/migrations/2024-11-25-203433_add_height_to_balances/up.sql:
```diff
@@ -0,0 +1,40 @@
+            OWNER,
+            token) max_heights ON bc.owner = max_heights.owner
+    AND bc.token = max_heights.token
+    AND bc.height = max_heights.max_height;
```
agree, if you could avoid having the migration and modify directly the balance table would be great! can keep the view, thanks!

joel-u410 replied:
sounds good, will do

joel-u410 replied:
done

joel-u410 commented on transactions/src/repository/transactions.rs:
```diff
@@ -18,6 +18,7 @@ pub fn insert_inner_transactions(
                 .map(InnerTransactionInsertDb::from)
                 .collect::<Vec<_>>(),
         )
+        .on_conflict_do_nothing()
```
I went ahead & removed that commit from this branch, since it is irrelevant to this feature. I can put it in a separate PR if it's still needed.

joel-u410 replied:
ok -- I remember now why I had done this!

If you ever want to force the indexer to re-index some past transactions, and manually update `last_processed_block` in the `crawler_state` where `name = 'transactions'` ... without the `on_conflict_do_nothing` you can put the indexer into an infinite loop.

Example:

First, update the `crawler_state` to "rewind history":
```
namada-indexer=# UPDATE crawler_state set last_processed_block = 38 where name = 'transactions';
UPDATE 1
```

Now, run `transactions`:
```
2024-12-06T18:02:07.412603Z  INFO transactions: Queried block successfully wrapper_txs=1 inner_txs=1 block=39
2024-12-06T18:02:07.423771Z ERROR shared::error: Database error reason=Failed to insert wrapper transactions in db

Caused by:
    duplicate key value violates unique constraint "wrapper_transactions_pkey"
2024-12-06T18:02:10.686714Z  INFO transactions: Queried block successfully wrapper_txs=1 inner_txs=1 block=39
2024-12-06T18:02:10.690671Z ERROR shared::error: Database error reason=Failed to insert wrapper transactions in db

Caused by:
    duplicate key value violates unique constraint "wrapper_transactions_pkey"
2024-12-06T18:02:14.555980Z  INFO transactions: Queried block successfully wrapper_txs=1 inner_txs=1 block=39
2024-12-06T18:02:14.564664Z ERROR shared::error: Database error reason=Failed to insert wrapper transactions in db

Caused by:
    duplicate key value violates unique constraint "wrapper_transactions_pkey"

... forever looping, never recovering ...
```

In this case, I'd rather it simply ignore the constraint violation (since already-indexed `transactions` should never change on-chain, we should not `UPDATE` them) and move on without failing.

I think this scenario can also occur if you were running two instances of the `transactions` crawler simultaneously (which you might want to do for HA purposes).

joel-u410 commented:
Updated the branch -- all comments addressed.

I also added a couple more commits here:

* Added some retry to `initial_query` because of occasional failure due to a race condition.
* Cleaned up the logging a bit, moved the more detailed logs into `debug` level.
